### PR TITLE
Add Teacache Support for LongCat Image

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
- Enables TeaCache support for LongCat Image. The model coefficients and speedups were calculated with the current config, not `main` (see [fix](https://github.com/vllm-project/vllm-omni/pull/1485)). 
- Includes some fixes to the coefficient estimator to avoid computing gradients and avoid dtype casting issues from running bf16 models
- Updates docs to add some notes on estimating coefficients for models that have layers that require vLLM's fwd context and parallel groups to be set up, since it was needed for this one 

### Example Outputs
For both text to image and image edit, teacache is the left one.
```bash
$ python text_to_image.py --cache-backend tea_cache --model meituan-longcat/LongCat-Image --output coffee_tc.png
$ python text_to_image.py --model meituan-longcat/LongCat-Image --output coffee.png
```

coffee_tc
coffee

For Image edit (using the coffee image above):
```bash
$ python image_edit.py --model meituan-longcat/LongCat-Image-Edit --image coffee.png --prompt "make the coffee cup transparent"  --cache-backend tea_cache --output edit_coffee_tc.png
$ python image_edit.py --model meituan-longcat/LongCat-Image-Edit --image coffee.png --prompt "make the coffee cup transparent" --output edit_coffee.png
```

edit_coffee_tc
edit_coffee

### Speed Benchmarks
With a thresh of .2, the speedup: is ~1.7x on an `h100`; didn't benchmark edit, but speedup looked comparable when I ran a quick check after.

Here is the full script I had used for testing, which can be used for reproduction for tti - it'll report the average speedup of 3 images

```python
import os
import gc
import time
import torch
from vllm_omni import Omni
from vllm_omni.inputs.data import OmniDiffusionSamplingParams

# Configuration
MODEL_ID = "meituan-longcat/LongCat-Image"
PROMPT = "A cup of coffee sitting on a table."
STEPS = 50
SEEDS = [444, 111, 3919]

TEACACHE_DIR = "cache_results"
NO_CACHE_DIR = "no_cache_results"
os.makedirs(TEACACHE_DIR, exist_ok=True)
os.makedirs(NO_CACHE_DIR, exist_ok=True)


def run_benchmark(use_cache=False):
    print(f"\n{'Testing with TeaCache' if use_cache else 'Testing without TeaCache'}...")
    times = []
    # Configure cache based on requirement
    out_dir = TEACACHE_DIR if use_cache else NO_CACHE_DIR
    cache_config = {
        "rel_l1_thresh": 0.2,
    } if use_cache else {}
    cache_backend = "tea_cache" if use_cache else None

    omni = Omni(
        model=MODEL_ID,
        cache_backend=cache_backend,
        cache_config=cache_config,
        dtype="bfloat16",
    )

    for seed in SEEDS:
        sampling_params = OmniDiffusionSamplingParams(num_inference_steps=STEPS, seed=seed)
        start = time.time()
        outputs = omni.generate(PROMPT, sampling_params)
        end = time.time()
        run_time = end - start
        times.append(run_time)
        # Save the generated image
        image = outputs[0].request_output[0].images[0]
        filename = f"{out_dir}/seed_{seed}.png"
        print(f"Run time: {run_time} for seed: {seed} [use_cache={use_cache}]")
        image.save(filename)

    avg_time = sum(times) / len(times)
    print(f"Average latency [use_cache={use_cache}]: {avg_time}")
    return avg_time


if __name__ == "__main__":
    # Run tests
    time_no_cache = run_benchmark(use_cache=False)
    torch.cuda.empty_cache()
    gc.collect()
    time_with_cache = run_benchmark(use_cache=True)

    print(f"\nResults:")
    print(f"Speedup: {time_no_cache / time_with_cache:.2f}x")
```

### CODE REVIEW

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 1]
```diff
@@ -566,6 +566,85 @@ def postprocess(h):
...
+    )
+    img_modulated = first_block.norm1(hs, emb=temb)[0]
+
+    # 3. Define the transformer execution
+    def run_transformer_blocks():
+        """Execute all Longcat transformer blocks."""
+        h = hidden_states
+        e = encoder_hidden_states
+        for block in module.transformer_blocks:
+            e, h = block(
```
[the remark this answers is not in the document]
Comment (alex-jw-brooks) [reply vllm_omni/diffusion/cache/teacache/extractors.py:793]: I agree that this should be optimized in the future - currently this behavior aligns with other teacache extractors though, so would be best to keep things consistent

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 2]
```diff
@@ -566,6 +566,85 @@ def postprocess(h):
...
+
+    temb = module.time_embed(timestep, hidden_states.dtype)
+    encoder_hidden_states = module.context_embedder(encoder_hidden_states)
+
+    ids = torch.cat((txt_ids, img_ids), dim=0)
+    image_rotary_emb = module.pos_embed(ids)
+
+    # 2. Extract the modulated output from the first mm-DiT block
+    first_block = module.transformer_blocks[0]
+    hs, _ = first_block(
```
[the remark this answers is not in the document]
Comment (alex-jw-brooks) [reply vllm_omni/diffusion/cache/teacache/extractors.py]: Good catch, will fix and rerun coefficient estimation 😬

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 3]
```diff
@@ -566,6 +566,85 @@ def postprocess(h):
...
+) -> CacheContext:
+    """"""
+    from diffusers.models.modeling_outputs import Transformer2DModelOutput
+
+    # 1. Model specific preprocessing
+    # TODO - fix sequence parallelism
+    sp_size = module.parallel_config.sequence_parallel_size
+    get_forward_context().sequence_parallel_size = sp_size
+
+    hidden_states = module.x_embedder(hidden_states)
```
[the remark this answers is not in the document]
Comment (alex-jw-brooks) [reply vllm_omni/diffusion/cache/teacache/extractors.py:764]: Useful, but I think there are larger underlying problems in SP for this model at the moment (see https://github.com/vllm-project/vllm-omni/issues/1556). I will investigate the fix for that as well, but see the same error with & without TeaCache at the moment, so open to any direction for how to handle it on this PR

File: vllm_omni/diffusion/cache/teacache/coefficient_estimator.py [hunk 4]
```diff
@@ -166,13 +190,17 @@ def __init__(
...
+        params = OmniDiffusionSamplingParams(
             num_inference_steps=generate_kwargs.get("num_inference_steps", 20),
             seed=generate_kwargs.get("seed", 42),
         )
-        self.pipeline.forward(req)
+        req = OmniDiffusionRequest(prompts=[prompt], sampling_params=params)
+        # Explicitly use inference mode to avoid gradients since we
+        # are not creating the pipeline through the model runner
+        with torch.inference_mode():
+            self.pipeline.forward(req)
```
Comment (alex-jw-brooks) [inline vllm_omni/diffusion/cache/teacache/coefficient_estimator.py:211]: A few small fixes were needed in this script to avoid OOMs in my env from gradients, and to handle bf16 since it can't be .numpy the tensors

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 5]
```diff
@@ -566,6 +574,144 @@ def postprocess(h):
...
+            torch.cat([txt_freqs_cos, img_freqs_cos], dim=0),
+            torch.cat([txt_freqs_sin, img_freqs_sin], dim=0),
+        )
+
+    # 2. Extract the modulated output from the first mm-DiT block
+    first_block = module.transformer_blocks[0]
+    _, hs = first_block(
+        hidden_states=hidden_states,
+        encoder_hidden_states=encoder_hidden_states,
+        temb=temb,
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/extractors.py]: This runs `first_block(...)` to get the modulated input, but then `run_transformer_blocks()` below iterates over all `module.transformer_blocks` (including `[0]`) again. So block 0 gets executed twice on every non-cached step.

The other extractors (e.g., qwen) avoid this by extracting the modulated input with just the lightweight norm call (`block.img_mod(temb)` + `block.img_norm1(...)`) without running the full block forward. Could you do something similar here, or at least start `run_transformer_blocks` from `module.transformer_blocks[1:]`?

File: vllm_omni/diffusion/cache/teacache/coefficient_estimator.py [hunk 6]
```diff
@@ -99,8 +103,27 @@ def install_hook(transformer: Any, hook: DataCollectionHook) -> None:
...
+        od_config = OmniDiffusionConfig.from_kwargs(model=model_path, dtype=dtype)
+        od_config.model_class_name = "LongCatImagePipeline"
+
+        pipeline = LongCatImagePipeline(od_config=od_config)
+        loader = DiffusersPipelineLoader(LoadConfig())
+        loader.load_weights(pipeline)
+        pipeline.to(device)
+        return pipeline
+
+
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/coefficient_estimator.py:134]: Should this also be wrapped with `set_default_torch_dtype(od_config.dtype)` like `BagelAdapter.load_pipeline` was updated to do above?

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 7]
```diff
@@ -566,6 +574,144 @@ def postprocess(h):
...
+    encoder_hidden_states,
+    txt_ids,
+    img_ids,
+    **kwargs,
+) -> CacheContext:
+    """"""
+    from diffusers.models.modeling_outputs import Transformer2DModelOutput
+    # TODO (Alex) - Refactor TeaCache extractors to more tightly integrate with .forward
+
+    # 1. Model specific preprocessing
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/extractors.py:754]: Nit: empty docstring — either remove it or add something minimal.

**Review by lishunyang12 — COMMENTED** [review]
Left a couple comments. The extractor mostly mirrors the model forward correctly, but the first block runs twice on non-cached steps which seems unintentional.

**Comment by hsliuustc0106** [comment]
Hi @alex-jw-brooks 👋

Checking in on the Teacache support for LongCat Image PR — 12 days since last update. Any progress?

Thanks!

**Comment by lishunyang12** [comment]
Hey @alex-jw-brooks — following up on the open threads from 2 weeks ago. The main concern is still the block 0 double execution + modulated input extraction.

Looking more carefully: `first_block.norm1(hs, emb=temb)[0]` extracts the modulated input from `hs` (the post-block-0 output), but it should be from the pre-block `hidden_states`. The Qwen extractor does this correctly — it calls `block.img_norm1(hidden_states, img_mod1)` on the *original* hidden states without running the full block. This means the cache decisions here are based on the wrong signal, and the coefficients were estimated with that bug.

Could you take a look?

**Comment by alex-jw-brooks** [comment]
Hey @hsliuustc0106 @lishunyang12, haven't forgotten about this PR, just paused it for a bit while fixing the sequence parallelism for this model to avoid copying things over here. I'll get back to it this afternoon and work on the comments, thanks for your patience 🙂

File: docs/user_guide/diffusion_acceleration.md [hunk 8]
```diff
@@ -55,8 +55,8 @@ The following table shows which models are currently supported by each accelerat
 
 | Model | Model Identifier | TeaCache | Cache-DiT | Ulysses-SP | Ring-Attention | CFG-Parallel | Tensor-Parallel | VAE-Patch-Parallel |
 |-------|------------------|:--------:|:---------:|:----------:|:--------------:|:------------:|:---------------:|:------------------:|
-| **LongCat-Image** | `meituan-longcat/LongCat-Image` | ❌ | ✅ | ❌ | ❌ | ✅ | ✅ | ❌ |
-| **LongCat-Image-Edit** | `meituan-longcat/LongCat-Image-Edit` | ❌ | ✅ | ❌ | ❌ | ✅ | ✅ | ❌ |
+| **LongCat-Image** | `meituan-longcat/LongCat-Image` | ✅  | ✅ | ✅  | ✅ | ✅ | ✅ | ❌ |
+| **LongCat-Image-Edit** | `meituan-longcat/LongCat-Image-Edit` | ✅  | ✅ | ✅  | ✅  | ✅ | ✅ | ❌ |
```
Comment (alex-jw-brooks) [inline docs/user_guide/diffusion_acceleration.md]: Columns for sequence parallel are out of date for LongCat; it does support ring attention and Ulysses SP.

Tested that teacache works with both types of SP on as part of this PR.

File: vllm_omni/diffusion/cache/teacache/extractors.py (hunk 7 above)
Comment (alex-jw-brooks) [reply vllm_omni/diffusion/cache/teacache/extractors.py:754]: Added 🙂

File: vllm_omni/diffusion/cache/teacache/extractors.py (hunk 5 above)
Comment (alex-jw-brooks) [reply vllm_omni/diffusion/cache/teacache/extractors.py]: Good catch 😬  Thanks! Fixed the modulated input and reran the coefficient calculations

File: vllm_omni/diffusion/cache/teacache/coefficient_estimator.py (hunk 6 above)
Comment (alex-jw-brooks) [reply vllm_omni/diffusion/cache/teacache/coefficient_estimator.py:134]: I had actually added a `set_default_torch_dtype` around the call to the load pipeline on the adapter instead of just putting it around the one line 🙂 the better way to do this is
```
        loader = DiffusersPipelineLoader(LoadConfig(), od_config=od_config)
        return loader.load_model(od_config=od_config, load_device=device)
```
because `load_model` will handle the device placement, put the model in eval mode, and handle the dtypes from the diffusion config. Updated both to avoid managing default dtypes manually and made sure the bagel one still runs

**Comment by alex-jw-brooks** [comment]
Hey @lishunyang12 @hsliuustc0106 thanks for the reviews - took another pass at this and added some additional info since I hadn't tested image edit yet originally. Ready for another look when you've got the bandwidth 🙂

**Comment by lishunyang12** [comment]
Thanks for the update. Will re-review this week.

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 9]
```diff
@@ -566,6 +569,101 @@ def postprocess(h):
...
+    modulated input extraction, transformer execution, and postprocessing
+    logic.
+
+    Args & kawrgs are identical to the inputs to LongCat Image's forward.
+
+    Returns:
+        CacheContext with all information needed for generic caching
+    """
+    # TODO (Alex) - Refactor TeaCache extractors to more tightly integrate with .forward
+    from diffusers.models.modeling_outputs import Transformer2DModelOutput
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/extractors.py:752]: This mutates `fwd_context.split_text_embed_in_sp` but never restores it. If the forward context is reused across timesteps, this side effect persists. Worth a comment or a reset in `postprocess()`.

**Review by lishunyang12 — APPROVED** [review]
The extractor looks correct now — norm1 is called on the pre-block hidden_states, and all blocks run in run_transformer_blocks(). Previous concern is addressed.

File: vllm_omni/diffusion/cache/teacache/extractors.py (hunk 9 above)
Comment (alex-jw-brooks) [reply vllm_omni/diffusion/cache/teacache/extractors.py:752]: Sure, I added a comment. For this model, we don't need to restore it because it should never be True, but IMO we should just remove this attribute from the repo for now, because I don't think it's currently ever expected to be `True` or even implemented for the `True` behavior. Will open a pr to discuss 🙂

**Comment by alex-jw-brooks** [comment]
@wtomin @hsliuustc0106 could you please review this PR?

I've refactored a bit to share some of the code for the estimators since LongCat / the new Stable Audio coefficient were basically the same, but it is ready for a look when you have a moment

**Comment by alex-jw-brooks** [comment]
Hi @lishunyang12 @hsliuustc0106 @wtomin, could you please take a look at this one when you have the chance?

**Comment by hsliuustc0106** [comment]
resolve conflicts please, I will add ready label for this PR

**Comment by alex-jw-brooks** [comment]
Thanks @hsliuustc0106! Just rebased 😄
